File: 0121_Best_Time_Buy_Sell_Stock.py
from typing import List
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxReturn = 0
        localMax = 0
        for i in range(1, len(prices)):
            localMax = max(0, localMax + prices[i] - prices[i-1])
            maxReturn = max(localMax, maxReturn)

        return maxReturn

File: 0121_Best_time_buy_sell_stock.cpp
#include <vector>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        int minBuy = prices[0];
        int maxProf = 0;

        for (std::vector<int>::iterator it = prices.begin(); it != prices.end(); it++) {
            if (*it < minBuy) {
                minBuy = *it;
            }
            else if (*it - minBuy > maxProf) {
                maxProf = *it - minBuy;
            }
        }

        return maxProf;
    }
};
